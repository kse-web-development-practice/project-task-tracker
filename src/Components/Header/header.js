import React, { useContext } from 'react'
import PropTypes from 'prop-types'
import styles from './header.module.css'
import { Button } from '../Button/button'
import { useNavigate } from 'react-router-dom'
import { UserContext } from '../../user-context'
import logo from './logo.png'

export const Header = ({ pageName, isAuthenticated }) => {
  const userContext = useContext(UserContext)
  const navigate = useNavigate()

  const onLogoClick = () => {
    navigate('/')
  }
  const onLoginClick = () => {
    navigate('/login')
  }
  const onSignupClick = () => {
    navigate('/register')
  }
  const onLogoutClick = () => {
    userContext.logout()
    navigate('/')
  }

  return (
    <header className={styles.header}>
      <div className={styles.logo} onClick={onLogoClick}>
        <img src={logo} alt="Logo" />
      </div>
      <div className={styles.projectName}>{pageName || 'Overview'}</div>

      {isAuthenticated ? (
        <Button onClick={onLogoutClick}>Logout</Button>
      ) : (
        <div className={styles.auth}>
          <Button onClick={onLoginClick}>Log in</Button>
          <Button onClick={onSignupClick} isMain={true}>
            Sign up
          </Button>
        </div>
      )}
    </header>
  )
}

Header.propTypes = {
  pageName: PropTypes.string,
  isAuthenticated: PropTypes.bool.isRequired
}
