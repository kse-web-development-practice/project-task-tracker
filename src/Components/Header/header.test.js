import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { Header } from './header'
import { UserContext } from '../../user-context'
import { BrowserRouter as Router } from 'react-router-dom'

const mockNavigate = jest.fn()

const mockLogout = jest.fn()

jest.mock('./logo.png', () => 'test-file-stub')

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockNavigate
}))

describe('Header Component', () => {
  const renderHeader = (isAuthenticated, pageName) =>
    render(
      <UserContext.Provider value={{ logout: mockLogout }}>
        <Router>
          <Header isAuthenticated={isAuthenticated} pageName={pageName} />
        </Router>
      </UserContext.Provider>
    )

  test('authenticated snapshot', () => {
    const { asFragment } = renderHeader(true, 'Hi, user!')
    expect(asFragment()).toMatchSnapshot()
  })

  test('not authenticated snapshot', () => {
    const { asFragment } = renderHeader(false)
    expect(asFragment()).toMatchSnapshot()
  })

  test('not authenticated button clicks', () => {
    renderHeader(false)

    fireEvent.click(screen.getByAltText('Logo'))
    expect(mockNavigate).toHaveBeenCalledWith('/')

    fireEvent.click(screen.getByText('Log in'))
    expect(mockNavigate).toHaveBeenCalledWith('/login')

    fireEvent.click(screen.getByText('Sign up'))
    expect(mockNavigate).toHaveBeenCalledWith('/register')
  })

  test('authenticated button clicks', () => {
    renderHeader(true)

    fireEvent.click(screen.getByAltText('Logo'))
    expect(mockNavigate).toHaveBeenCalledWith('/')

    fireEvent.click(screen.getByText('Logout'))
    expect(mockLogout).toHaveBeenCalled()
  })
})
