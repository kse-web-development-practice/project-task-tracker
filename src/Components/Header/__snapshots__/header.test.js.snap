// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component authenticated snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="logo"
    >
      <img
        alt="Logo"
        src="test-file-stub"
      />
    </div>
    <div
      class="projectName"
    >
      Hi, user!
    </div>
    <button
      class="button "
    >
      Logout
    </button>
  </header>
</DocumentFragment>
`;

exports[`Header Component not authenticated snapshot 1`] = `
<DocumentFragment>
  <header
    class="header"
  >
    <div
      class="logo"
    >
      <img
        alt="Logo"
        src="test-file-stub"
      />
    </div>
    <div
      class="projectName"
    >
      Overview
    </div>
    <div
      class="auth"
    >
      <button
        class="button "
      >
        Log in
      </button>
      <button
        class="button main"
      >
        Sign up
      </button>
    </div>
  </header>
</DocumentFragment>
`;
